Provide PHP 8.1 support

weierophinney opened the pull request:
This patch provides PHP 8.1 support for laminas-mail via the following changes:

- Adds `~8.1.0` to the list of allowed PHP versions
- Bumps dependencies to versions known to work with 8.0/8.1 where possible
- Removes laminas-config as a dependency
  - It was a soft dependency previously; the source code did not directly reference it.
  - Refactoring of storage adapters was done to normalize options/params to associative arrays; this will also work on laminas-config instances, providing backwards compatibility.
- Changes how the package replaces zend-mail
  - Renames "replace" section to "conflict"
  - Changes "zend-mail" constraint in that section to "*"
  - Removes dependency on laminas-zendframework-bridge
- Adds `#[ReturnTypeWillChange]` attributes where needed
- When ext-intl is present, we now use `idn_to_ascii()` for punycode conversion, as the true/punycode package raises errors due to "implicit float to int conversion" under 8.1, leading to broken punycode conversion.
- `list() = explode()` raises a warning under 8.1 if an index is not created, vs a notice in previous versions; the patch ~~now detects this~~ refactors these to test for delimiters before attempting to explode a string.

weierophinney committed:
feat: prepare for PHP 8.1 support

- Add `~8.1.0` to list of allowed PHP versions
- Bump dependencies to versions known to work with 8.0/8.1 where possible
- Remove laminas-config as a dependency (never referenced other than in docblocks)
- Change how the package replaces zend-mail
  - Rename "replace" section to "conflict"
  - Change "zend-mail" constraint in that section to "*"
  - Remove dependency on laminas-zendframework-bridge

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: remove laminas-config requirement in tests

The various storage adapters are really only looking for maps.
This patch adds a trait, `Laminas\Mail\Storage\ParamsNormalizerTrait`, which provides functionality for producing an array map from any of a `Traversable` object, a standard object, or an array (validating that only string keys are provided).
Each storage adapter constructor has been updated to use this and to then validate the values pulled from the array.
When done, there were 4 new `MixedAssignment` errors that needed to be added to the Psalm baseline.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
feat: add #[ReturnTypeWillChange] attributes

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: provide fixes for PHP 8.1

- Bump laminas-mime to 2.9.1 to pick up a fix in its `Decode` class
- Provide alternative punycode conversion via ext-intl `idn_to_ascii()` function when present. true/punycode has an "implicit float to int conversion` error under 8.1 that leads to valid punycode values raising errors and not being converted.
- Ensure `GenericHeader` has string values for each of the field name and value.
  This is what they were already documented as being, but there was no verification, and contradictory behavior/documentation.
- `list() = explode()` raises a warning now if an index is not created, vs a notic.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: update exception expectations for invalid/missing maildir

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: ignore platform PHP requirement when testing under PHP 8.1

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: undefined array key is now E_WARNING, not E_NOTICE, when list dereferencing

Found under PHP 8.0

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: check for string before passing value to ctype_digit

Passing non-string value to ctype_digit is deprecated starting in 8.1

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: null coalesce to string before passing value to explode()

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Ocramius commented on src/Storage/Folder/Mbox.php:
```diff
@@ -125,7 +133,7 @@ public function getFolders($rootFolder = null)
         $subname = trim($rootFolder, DIRECTORY_SEPARATOR);
         while ($currentFolder) {
-            ErrorHandler::start(E_NOTICE);
+            ErrorHandler::start(E_NOTICE | E_WARNING);
```
An error handler used around an `explode()` call? O_o

Ocramius replied:
Also, the added `E_WARNING` seems a bit much - perhaps we really need to just check `$subname` before calling `explode()` ?

Ocramius commented on src/Storage/ParamsNormalizerTrait.php:
```diff
@@ -0,0 +1,34 @@
+use Traversable;
+use Webmozart\Assert\Assert;
+
+trait ParamsNormalizerTrait
```
Should be `/** @internal */`

Ocramius commented on src/Storage/ParamsNormalizerTrait.php:
```diff
@@ -0,0 +1,34 @@
+     * @param mixed $params
+     * @return array<string, mixed>
+     */
+    private function normalizeParams($params): array
```
Let's not use a trait for this: a `public static function` works just as fine, without the complexities of traits

Ocramius commented on src/Storage/Maildir.php:
```diff
@@ -295,14 +301,14 @@ protected function getMaildirFiles($dh, $dirname, $defaultFlags = [])
             }
 
-            ErrorHandler::start(E_NOTICE);
+            ErrorHandler::start(E_NOTICE | E_WARNING);
```
Similar to above: a whole `ErrorHandler` usage just because we operate with `explode()` on unknown input (or assume it will produce less than 2 results)? :thinking:

Ocramius commented on src/Storage/ParamsNormalizerTrait.php:
```diff
@@ -0,0 +1,34 @@
+     */
+    private function normalizeParams($params): array
+    {
+        if ($params instanceof Traversable) {
```
This logic needs a unit test, once converted into a `static` method

weierophinney committed:
qa: switch ParamsNormalizer from trait to final static class

Renames `ParamsNormalizerTrait` to `ParamsNormalizer`, making it a final class with a single static method.
Adds unit tests for the method, and updates code that referenced the trait previously to reference the static method.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
refactor: remove error handling when related to explode/list operations

Per review, usage of `ErrorHandler::start()/stop()` around explode/list operations is brittle and makes understanding what types are assigned.
Refactored these to use conditionals, checking for delimiters in strings before attempting such operations.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Ocramius reviewed (approved):
LGTM 👍
